### PR TITLE
Network resource - Trunk

### PR DESCRIPTION
**Pull request by celebdor**
This patch support for the OpenStack networking trunk extension.

For #345 

Signed-off-by: Antoni Segura Puimedon <celebdor@gmail.com>

### CODE REVIEW

File: openstack/resource_openstack_compute_instance_v2_test.go [hunk 1]
```diff
@@ -724,6 +725,36 @@ func testAccCheckComputeV2InstanceState(
 	}
 }
 
+func TestAccComputeV2Instance_trunk(t *testing.T) {
```
Comment (jtopjian) [inline openstack/resource_openstack_compute_instance_v2_test.go]: Since the "trunk" resource is the core focus of this test, let's move this to a dedicated `resource_openstack_networking_trunk_v2_test.go` file.

File: openstack/resource_openstack_networking_trunk_v2.go [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: false,
+			},
+			"port_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"admin_state_up": &schema.Schema{
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_trunk_v2.go]: Let's change this to look like: https://github.com/terraform-providers/terraform-provider-openstack/blob/e8fa32f8549b144625f0a4b011be5f83eb0aae2f/openstack/data_source_openstack_networking_router_v2.go#L29-L32

And then access it with: https://github.com/terraform-providers/terraform-provider-openstack/blob/e8fa32f8549b144625f0a4b011be5f83eb0aae2f/openstack/data_source_openstack_networking_router_v2.go#L93-L96

There are some `admin_state_up` arguments in other resources that are `TypeString`. Those were some of the first resources before Terraform was able to handle booleans better.

File: openstack/resource_openstack_networking_trunk_v2.go [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+						},
+						"segmentation_id": &schema.Schema{
+							Type:     schema.TypeInt,
+							Required: true,
+						},
+					},
+				},
+			},
+			"tags": &schema.Schema{
+				Type:     schema.TypeSet,
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_trunk_v2.go]: This might work better as a `TypeMap`?

https://github.com/terraform-providers/terraform-provider-openstack/blob/e9fe36a257561b7b0a7e29ff409e504c8305facf/openstack/resource_openstack_compute_flavor_v2.go#L69-L73

File: openstack/resource_openstack_networking_trunk_v2.go [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+	stateConf := &resource.StateChangeConf{
+		Target:     []string{"ACTIVE"},
+		Refresh:    waitForNetworkTrunkActive(client, trunk.ID),
+		Timeout:    d.Timeout(schema.TimeoutCreate),
+		Delay:      5 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_trunk_v2.go]: Let's make sure we check for an error here:

```go
if err != nil {
  return fmt.Errorf("Error waiting for trunk to become active: %s", err)
}
```

File: openstack/resource_openstack_networking_trunk_v2.go [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+	d.Set("tags", trunk.Tags)
+
+	subports := make([]map[string]interface{}, len(trunk.Subports))
+	for i, trunkSubport := range trunk.Subports {
+		subports[i] = make(map[string]interface{})
+		subports[i]["port_id"] = trunkSubport.PortID
+		subports[i]["segmentation_type"] = trunkSubport.SegmentationType
+		subports[i]["segmentation_id"] = trunkSubport.SegmentationID
+	}
+	d.Set("sub_ports", subports)
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_trunk_v2.go]: For "non-scalar" types, it's best to check for an error:

```go
if err := d.Set("sub_ports", subports); err != nil {
  return fmt.Errorf("Unable to set sub_port for trunk %s: %s", d.Id(), err)
}
```

File: website/docs/r/networking_trunk_v2.html.markdown [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    (must be "true" or "false" if provided). Changing this updates the
+    `admin_state_up` of an existing trunk.
+
+* `tenant_id` - (Optional) The owner of the Trunk. Required if admin wants
+    to create a trunk on behalf of another tenant. Changing this creates a new trunk.
+
+* `sub_ports` - (Optional) The set of ports that will be made subports of the trunk.
+    The structure of each subport is described below.
+
+* `value_specs` - (Optional) Map of additional options.
```
Comment (jtopjian) [inline website/docs/r/networking_trunk_v2.html.markdown]: `value_specs` isn't defined here, so let's remove this entry

File: website/docs/r/networking_trunk_v2.html.markdown [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+The `subports` block supports:
```
Comment (jtopjian) [inline website/docs/r/networking_trunk_v2.html.markdown]: `sub_port`

File: website/docs/r/networking_trunk_v2.html.markdown [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+## Attributes Reference
+
+The following attributes are exported:
+
+* `region` - See Argument Reference above.
+* `name` - See Argument Reference above.
+* `port_id` - See Argument Reference above.
+* `admin_state_up` - See Argument Reference above.
+* `tenant_id` - See Argument Reference above.
+* `subports` - See Argument Reference above
```
Comment (jtopjian) [inline website/docs/r/networking_trunk_v2.html.markdown]: `sub_port`

File: openstack/resource_openstack_networking_trunk_v2.go [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+				ForceNew: false,
+				Computed: true,
+			},
+			"tenant_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Computed: true,
+			},
+			"sub_ports": &schema.Schema{
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_trunk_v2.go]: This might read better if this was called `sub_port` instead of `sub_ports`:

```hcl
resource "openstack_networking_trunk_v2" "foo" {
  ...
  sub_port {
    segmentation_id = 1
  }

  sub_port {
    segmentation_id = 2
  }
}
```

File: openstack/resource_openstack_networking_trunk_v2.go [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+		Delay:      5 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error deleting OpenStack Neutron trunk: %s", err)
+	}
+
+	d.SetId("")
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_trunk_v2.go]: nit: This is no longer needed. It used to be in older versions of Terraform and might be hanging around some resources.

File: openstack/resource_openstack_networking_trunk_v2.go [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+	return nil
+}
+
+func resourceSubportsMap(d *schema.Set) (subports map[trunks.Subport]bool) {
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_trunk_v2.go]: nit: let's make this function name more specific to the resource. Something like:

`resourceNetworkingV2TrunkSubportsMap`

Same for the other function names, too. It helps prevent collisions under a single namespace.

File: openstack/resource_openstack_compute_instance_v2_test.go (hunk 1 above)
Comment (celebdor) [reply openstack/resource_openstack_compute_instance_v2_test.go]: Done. Thanks.

File: openstack/resource_openstack_networking_trunk_v2.go (hunk 2 above)
Comment (celebdor) [reply openstack/resource_openstack_networking_trunk_v2.go]: Done

File: openstack/resource_openstack_networking_trunk_v2.go (hunk 9 above)
Comment (celebdor) [reply openstack/resource_openstack_networking_trunk_v2.go]: Done, thanks.

File: openstack/resource_openstack_networking_trunk_v2.go (hunk 3 above)
Comment (celebdor) [reply openstack/resource_openstack_networking_trunk_v2.go]: Tags are not key value like:
```  labels {
    kube_tag                         = "1.11.1"
    prometheus_monitoring            = "true"
    influx_grafana_dashboard_enabled = "true"
    kube_dashboard_enabled           = "true"
  }
```

Rather, they are like:
```
{
   "port":{
      "status":"DOWN",
      "created_at":"2018-10-16T20:25:12Z",
      "description":"",
      "allowed_address_pairs":[

      ],
      "tags":[
         "bar",
         "foo"
      ],
      "network_id":"360451b6-d2d2-4b9d-b1cb-f6fe8254103c",
      "tenant_id":"e153f3f9082240a5974f667cfe1036e3",
      "extra_dhcp_opts":[

      ],
      "admin_state_up":true,
      "updated_at":"2018-10-17T09:11:26Z",
      "name":"subport_3",
      "device_owner":"",
      "revision_number":3,
      "mac_address":"fa:16:3e:29:2b:be",
      "binding:vnic_type":"normal",
      "port_security_enabled":true,
      "project_id":"e153f3f9082240a5974f667cfe1036e3",
      "fixed_ips":[

      ],
      "id":"f1d082f2-aa3d-4955-a400-ceea7c5ffc28",
      "security_groups":[
         "ca55e00b-75dd-4292-b76d-977e133f9689"
      ],
      "device_id":""
   }
}```

File: openstack/resource_openstack_networking_trunk_v2.go (hunk 5 above)
Comment (celebdor) [reply openstack/resource_openstack_networking_trunk_v2.go]: Great idea!

File: openstack/resource_openstack_networking_trunk_v2.go (hunk 10 above)
Comment (celebdor) [reply openstack/resource_openstack_networking_trunk_v2.go]: Great. Thanks

File: website/docs/r/networking_trunk_v2.html.markdown (hunk 7 above)
Comment (celebdor) [reply website/docs/r/networking_trunk_v2.html.markdown]: Done!

File: website/docs/r/networking_trunk_v2.html.markdown (hunk 8 above)
Comment (celebdor) [reply website/docs/r/networking_trunk_v2.html.markdown]: Done!

File: website/docs/r/networking_trunk_v2.html.markdown (hunk 6 above)
Comment (celebdor) [reply website/docs/r/networking_trunk_v2.html.markdown]: Done!

**Comment by jtopjian** [comment]
> Acceptance testing should catch any side effects.

And sure enough, unit and acceptance testing caught a problem. I recently merged https://github.com/gophercloud/gophercloud/pull/1294 so I can fix a Terraform issue. We'll have to get that issue fixed before this PR can be merged. We should be able to have that done in the next day or two.

**Comment by celebdor** [comment]
I saw compilation errors that needed this unrelated change:

```
--- a/openstack/resource_openstack_lb_member_v2.go
+++ b/openstack/resource_openstack_lb_member_v2.go
@@ -99,12 +99,13 @@ func resourceMemberV2Create(d *schema.ResourceData, meta interface{}) error {
        }
 
        adminStateUp := d.Get("admin_state_up").(bool)
+        weight := d.Get("weight").(int)
        createOpts := pools.CreateMemberOpts{
                Name:         d.Get("name").(string),
                TenantID:     d.Get("tenant_id").(string),
                Address:      d.Get("address").(string),
                ProtocolPort: d.Get("protocol_port").(int),
-               Weight:       d.Get("weight").(int),
+               Weight:       &weight,
                AdminStateUp: &adminStateUp,
        }
 
@@ -186,7 +187,8 @@ func resourceMemberV2Update(d *schema.ResourceData, meta interface{}) error {
                updateOpts.Name = d.Get("name").(string)
        }
        if d.HasChange("weight") {
-               updateOpts.Weight = d.Get("weight").(int)
+               weight := d.Get("weight").(int)
+               updateOpts.Weight = &weight
        }
        if d.HasChange("admin_state_up") {
                asu := d.Get("admin_state_up").(bool)
```

**Comment by jtopjian** [comment]
> I saw compilation errors that needed this unrelated change:

Right. It's best to not bundle this change into this PR. I'll have a separate PR shortly.

The reason for the upstream change is for #392

**Comment by celebdor** [comment]
Cool. Thanks!

**Comment by celebdor** [comment]
@jtopjian In the meantime, should I add support for setting tags for the trunk resource, now that gophercloud supports it, or is it better to make a second PR after this one is merged?

**Comment by jtopjian** [comment]
@celebdor sorry for the delay. Since this is a brand new resource that hasn't been merged, sure, bring in the new Gophercloud code and add Tags :)

If this was an existing resource and you were adding two features, then 2 PRs would be better. But that's not the case here.

File: openstack/resource_openstack_networking_trunk_v2.go (hunk 3 above)
Comment (jtopjian) [reply openstack/resource_openstack_networking_trunk_v2.go]: Oh duh. I'm sorry - I was mistaking tags for metadata.

File: openstack/resource_openstack_networking_trunk_v2.go [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+	value := false
+
+	if v, ok := d.GetOkExists("admin_state_up"); ok {
+		value = v.(bool)
+	}
+
+	return &value
+}
+
+func resourceTags(d *schema.ResourceData) (tags []string) {
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_trunk_v2.go]: It's very possible this function can be used for any of the future network tags field, so how about we move it to the `util.go` file? Let's call it `networkV2AttributesTags` since there could be another tag implementation in another OpenStack project that does tags differently.

Comment (hardys) [reply openstack/resource_openstack_networking_trunk_v2.go]: Agreed this will be useful for other resources - I have a WIP branch at https://github.com/hardys/terraform-provider-openstack/tree/neutron_tags which does as suggested by @jtopjian but since @celebdor beat me to it, I'll rebase on this PR :)

File: openstack/resource_openstack_networking_trunk_v2.go [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+		return CheckDeleted(d, err, "trunk")
+	}
+	log.Printf("[DEBUG] Retrieved trunk %s: %+v", d.Id(), trunk)
+
+	d.Set("region", GetRegion(d, config))
+	d.Set("name", trunk.Name)
+	d.Set("port_id", trunk.PortID)
+	d.Set("admin_state_up", trunk.AdminStateUp)
+	d.Set("tenant_id", trunk.TenantID)
+	d.Set("tags", trunk.Tags)
```
Comment (hardys) [inline openstack/resource_openstack_networking_trunk_v2.go]: Perhaps I missed it, but did you add any test proving the set/get/update of tags?

If we're adding tags in this commit it'd be nice to test them (particularly since I've run into a few issues with the acceptance test locally and I'd like to confirm if it'd my code or a problem that will also occur with this patch).

**Comment by celebdor** [comment]
I'm pushing the test for flags now

**Comment by celebdor** [comment]
@jtopjian @hardys Note that I also added an acceptance check in utils.go that I think could be used for the other resources.

**Comment by jtopjian** [comment]
@celebdor Separate from the review I just submitted, after reading through this resource and running the acceptance tests, I'm wondering if we should break this into two separate resources:

* `openstack_networking_trunk_v2`
* `openstack_networking_trunk_subport_v2`

So an implementation would look like:

```hcl
resource "openstack_networking_trunk_v2" "trunk_1" {          
  name = "trunk_1"                                            
  port_id = "${openstack_networking_port_v2.parent_port_1.id}"
  admin_state_up = "true"                                     
}                                                             

resource "openstack_networking_trunk_subport_v2" "subport_1" {
  trunk_id = "${openstack_networking_trunk_v2.trunk_1.id}"
  port_id = "${openstack_networking_port_v2.subport_1.id}"
  segmentation_id = 1
  segmentation_type = "vlan"
}

resource "openstack_networking_trunk_subport_v2" "subport_2" {
  trunk_id = "${openstack_networking_trunk_v2.trunk_1.id}"
  port_id = "${openstack_networking_port_v2.subport_2.id}"
  segmentation_id = 2
  segmentation_type = "vlan"
}
```

The disadvantage is that it's more verbose and that you won't be able to add subports during the initial creation of the trunk.

However, the advantage is that a separate resource is more flexible: users can add and remove subports using Terraform's native HCL logic and users can more easily reference individual subports. For example, here's what 3 subports look like in the Terraform state:

```
sub_port.# = 3
sub_port.1580179709.port_id = b5689463-1426-44d6-a8fc-1efe56ac389b
sub_port.1580179709.segmentation_id = 1
sub_port.1580179709.segmentation_type = vlan
sub_port.272728530.port_id = 9852a0a1-b47a-4730-9d23-0d0df1fba9a7
sub_port.272728530.segmentation_id = 3
sub_port.272728530.segmentation_type = vlan
sub_port.504626688.port_id = f304bf3b-175d-4d5a-b68a-e873136b96b2
sub_port.504626688.segmentation_id = 4
sub_port.504626688.segmentation_type = vlan
```

The index is a computed hash that is not predictable and makes it near impossible for a user to reference it with `${openstack_networking_trunk_v2.sub_ports.123456.segmentation_type}`.

Historically, any time we have had a "sub-resource" like this in a parent resource, users end up using all sorts of hacks to better manage the sub-resource. Two good examples are:

* https://github.com/terraform-providers/terraform-provider-openstack/issues/170
* https://github.com/terraform-providers/terraform-provider-openstack/issues/174

Terraform 0.12 _should_ provide a way to more easily handle these situations, but if the resources were designed to be more "composable" from the start, there wouldn't be an issue. As an aside, doing a _very_ large refactor of the `openstack_networking_port_v2` resource is on our TODO list to remedy this exact situation :)

Thoughts?

File: openstack/resource_openstack_networking_trunk_v2.go [hunk 14]
```diff
@@ -0,0 +1,354 @@
...
+		log.Printf("[DEBUG] Updating trunk %s with options: %+v", d.Id(), updateOpts)
+		_, err = trunks.Update(client, d.Id(), updateOpts).Extract()
+		if err != nil {
+			return fmt.Errorf("Error updating OpenStack Neutron trunk: %s", err)
+		}
+	}
+
+	var subportsToAdd trunks.AddSubportsOpts
+	var subportsToRemove trunks.RemoveSubportsOpts
+	if d.HasChange("sub_port") {
```
[the remark this answers is not in the document]
Comment (celebdor) [reply openstack/resource_openstack_networking_trunk_v2.go]: That's great! I was trying to find set operations and implemented this cause I didn't find them. Mush better this way!

File: openstack/resource_openstack_networking_trunk_v2.go (hunk 13 above)
Comment (celebdor) [reply openstack/resource_openstack_networking_trunk_v2.go]: There wasn't at the time, now it is:
TestAccNetworkingV2Trunk_tags

**Comment by celebdor** [comment]
> @celebdor Separate from the review I just submitted, after reading through this resource and running the acceptance tests, I'm wondering if we should break this into two separate resources:
> 
>     * `openstack_networking_trunk_v2`
> 
>     * `openstack_networking_trunk_subport_v2`
> 
> 
> So an implementation would look like:
> 
> ```terraform
> resource "openstack_networking_trunk_v2" "trunk_1" {          
>   name = "trunk_1"                                            
>   port_id = "${openstack_networking_port_v2.parent_port_1.id}"
>   admin_state_up = "true"                                     
> }                                                             
> 
> resource "openstack_networking_trunk_subport_v2" "subport_1" {
>   trunk_id = "${openstack_networking_trunk_v2.trunk_1.id}"
>   port_id = "${openstack_networking_port_v2.subport_1.id}"
>   segmentation_id = 1
>   segmentation_type = "vlan"
> }
> 
> resource "openstack_networking_trunk_subport_v2" "subport_2" {
>   trunk_id = "${openstack_networking_trunk_v2.trunk_1.id}"
>   port_id = "${openstack_networking_port_v2.subport_2.id}"
>   segmentation_id = 2
>   segmentation_type = "vlan"
> }
> ```
> 
> The disadvantage is that it's more verbose and that you won't be able to add subports during the initial creation of the trunk.
> 
> However, the advantage is that a separate resource is more flexible: users can add and remove subports using Terraform's native HCL logic. For example:
> 
> ```terraform
> resource "openstack_networking_trunk_v2" "trunk_1" {          
>   name           = "trunk_1"                                            
>   port_id        = "${openstack_networking_port_v2.parent_port_1.id}"
>   admin_state_up = "true"                                     
> }                                                             
> 
> resource "openstack_networking_trunk_subport_v2" "subports" {
>   count             = 3
>   trunk_id          = "${openstack_networking_trunk_v2.trunk_1.id}"
>   port_id           = "${openstack_networking_port_v2.ports.*.id[count.index]}"
>   segmentation_id   = "${4000 + count.index}"
>   segmentation_type = "vlan"
> }
> ```
> 
> and users can more easily reference individual subports. For example, here's what 3 subports look like in the Terraform state:
> 
> ```
> sub_port.# = 3
> sub_port.1580179709.port_id = b5689463-1426-44d6-a8fc-1efe56ac389b
> sub_port.1580179709.segmentation_id = 1
> sub_port.1580179709.segmentation_type = vlan
> sub_port.272728530.port_id = 9852a0a1-b47a-4730-9d23-0d0df1fba9a7
> sub_port.272728530.segmentation_id = 3
> sub_port.272728530.segmentation_type = vlan
> sub_port.504626688.port_id = f304bf3b-175d-4d5a-b68a-e873136b96b2
> sub_port.504626688.segmentation_id = 4
> sub_port.504626688.segmentation_type = vlan
> ```
> 
> The index is a computed hash that is not predictable and makes it near impossible for a user to reference it with `${openstack_networking_trunk_v2.sub_ports.123456.segmentation_type}`.
> 
> Historically, any time we have had a "sub-resource" like this in a parent resource, users end up using all sorts of hacks to better manage the sub-resource. Two good examples are:
> 
>     * #170
> 
>     * #174
> 
> 
> Terraform 0.12 _should_ provide a way to more easily handle these situations, but if the resources were designed to be more "composable" from the start, there wouldn't be an issue. As an aside, doing a _very_ large refactor of the `openstack_networking_port_v2` resource is on our TODO list to remedy this exact situation :)
> 
> Thoughts?

The HCL argument for using count on ports convinced me to give it a try. I'll try to get it done over the weekend and see how it looks like.

**Comment by celebdor** [comment]
> The HCL argument for using count on ports convinced me to give it a try. I'll try to get it done over the weekend and see how it looks like.

I'm worried about having a separate resource meaning that we'll need to add each subport one by one instead of being able to do it in a single add_subports remove_subports operation. This would add a significant load on Neutron.

> 
> Thanks a lot for the background info and all the improvement proposals. They are much, much appreciated!

File: openstack/resource_openstack_networking_trunk_v2.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (jtopjian) [reply openstack/resource_openstack_networking_trunk_v2.go]: I just realized that `len(toAdd.List())` can be changed to `toAdd.Len()`

**Comment by jtopjian** [comment]
> I'm worried about having a separate resource meaning that we'll need to add each subport one by one instead of being able to do it in a single add_subports remove_subports operation. This would add a significant load on Neutron.

Good point - this is something to consider.

There are 3 possible execution methods in play here:

1. One single resource bundles all subport changes into 1 API call

2. Two separate resources would cause all subport changes to happen _in parallel_ in separate API calls, each call performing an action on 1 subport. This is because Terraform is parallel in nature.  
  
    For the example with `count = 3`, 3 parallel `add_subports` calls would happen, one for each subport. In my opinion, this execution method will cause the most possible load/problems with Neutron since parallel requests hitting the ML2 driver to do the actual physical work on the NIC/vNIC might cause concurrency issues. 

    A similar existing example might be the `openstack_networking_secgroup_v2` and `openstack_networking_secgroup_rule_v2` resources. They execute in the same, parallel fashion and we haven't had any reported problems. That's not to say problems can't happen, though.

3. We can place a [mutex](https://golang.org/pkg/sync/) on the `openstack_networking_trunk_subport_v2` resource to _enforce_ it to run serial. This would be equivalent to a user doing `openstack network trunk set --subport` one by one.

The main reason I proposed a separate resource is to enable easier composition of trunks and subports using HCL. But if that is going to cause problems with the Neutron service itself, it might be a bad idea. As I mentioned, Terraform 0.12 should be able to handle [composition within a resource](https://www.hashicorp.com/blog/hashicorp-terraform-0-12-preview-for-and-for-each), so perhaps we can consider this a non-issue.

The second benefit of a separate resource is to get around the inability to reference a subport using the auto-generated index. Again, maybe Terraform 0.12 can [help with this](https://www.hashicorp.com/blog/terraform-0-12-generalized-splat-operator), but I'm not 100% sure yet. I'm OK with considering this a non-blocking issue at the moment.

so tl;dr: If there is serious concern about overloading the API and serializing the requests (method 3) isn't good, let's proceed with a single resource. The `sub_port` argument in `openstack_networking_trunk_v2` isn't "computed" and that will allow us an easy path of introducing a `openstack_networking_trunk_subport_v2` resource in the future if there's enough user feedback for it.

**Comment by celebdor** [comment]
I addressed the comments and fixed the build:

```
[stack@kuryr-a32c-02 openstack]$ TF_ACC=1 go test . -v --run=TestAccNetworkingV2Trunk_ -count=1 -timeout 120m
=== RUN   TestAccNetworkingV2Trunk_nosubports
--- PASS: TestAccNetworkingV2Trunk_nosubports (49.46s)
=== RUN   TestAccNetworkingV2Trunk_subports
--- PASS: TestAccNetworkingV2Trunk_subports (54.50s)
=== RUN   TestAccNetworkingV2Trunk_tags
--- PASS: TestAccNetworkingV2Trunk_tags (51.55s)
=== RUN   TestAccNetworkingV2Trunk_trunkUpdateSubports
--- PASS: TestAccNetworkingV2Trunk_trunkUpdateSubports (66.52s)
=== RUN   TestAccNetworkingV2Trunk_computeInstance
--- PASS: TestAccNetworkingV2Trunk_computeInstance (99.29s)
PASS
```

In regards to the subport options. After discussing it a bit with my team, we feel like:

* Option 2 is going to cause Neutron server load problems indeed, so we'd rather avoid that
* Option 3 with serialization via mutex would make creating big pools slow.
* Option 1 is missing count to be able to create trunk pools, but we hope that Terraform 0.12 will come soon and it will make the issue go away.

I suppose there is really no way to bundle resource creation for bulk creation/deletion in terraform (note that this would also be useful for the port resource, that supports bulk operations in the neutron api), is there?

File: website/openstack.erb [hunk 15]
```diff
@@ -67,6 +67,9 @@
             <li<%= sidebar_current("docs-openstack-datasource-networking-subnetpool-v2") %>>
               <a href="/docs/providers/openstack/d/networking_subnetpool_v2.html">openstack_networking_subnetpool_v2</a>
             </li>
+            <li<%= sidebar_current("docs-openstack-datasource-networking-trunk-v2") %>>
+              <a href="/docs/providers/openstack/d/networking_trunk_v2.html">openstack_networking_trunk_v2</a>
```
Comment (jtopjian) [inline website/openstack.erb]: This is linking to `/d/` which is for data sources. It'll need moved further below to the `/r/` sections. Travis will flag an error if the link is incorrect: https://travis-ci.org/terraform-providers/terraform-provider-openstack/builds/444878986#L554-L555

File: openstack/resource_openstack_networking_trunk_v2_test.go [hunk 16]
```diff
@@ -0,0 +1,690 @@
...
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/trunks"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/ports"
+)
+
+func TestAccNetworkingV2Trunk_nosubports(t *testing.T) {
+	var port_1 ports.Port
+	var trunk_1 trunks.Trunk
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
```
Comment (jtopjian) [inline openstack/resource_openstack_networking_trunk_v2_test.go]: Similar with Gophercloud, it looks like OpenLab's devstack configuration does not enable the trunk extension at this time, so acceptance tests are always going to fail. Normally in this situation, we add a new environment variable [here](https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/provider_test.go#L15) to toggle the valid type of configuration, but this seems excessive for the trunk extension. 

Instead, let's change this to an admin-only test (which won't run under Terraform / OpenLab at this time) and I will ask the OpenLab team to enable the trunk extension.

Let's change all `PreCheck` entries to:

```go
PreCheck: func() {           
  testAccPreCheck(t)         
  testAccPreCheckAdminOnly(t)
},                           
```

File: openstack/util.go [hunk 17]
```diff
@@ -187,3 +191,47 @@ func containerInfraLabelsStringV1(d *schema.ResourceData) (string, error) {
...
+	rawTags := d.Get("tags").(*schema.Set).List()
+	tags = make([]string, len(rawTags))
+
+	for i, raw := range rawTags {
+		tags[i] = raw.(string)
+	}
+	return
+}
+
+func testAccCheckNetworkingV2Tags(name string, key string, tags []string) resource.TestCheckFunc {
```
Comment (hardys) [inline openstack/util.go]: Small nit - we could probably remove the key parameter here, since it's always going to be "tags"?

Comment (celebdor) [reply openstack/util.go]: Sure, why not :+1:

File: openstack/resource_openstack_networking_trunk_v2_test.go (hunk 16 above)
Comment (celebdor) [reply openstack/resource_openstack_networking_trunk_v2_test.go]: Got it! Will do!

**Comment by celebdor** [comment]
We can merge. Thanks!
